diff -urNp matplotlib/numerix/ma.init/__init__.py matplotlib/numerix/ma/__init__.py
--- matplotlib/numerix/ma.init/__init__.py	2006-08-19 16:21:56.000000000 -0400
+++ matplotlib/numerix/ma/__init__.py	2006-11-29 12:48:14.000000000 -0500
@@ -9,7 +9,9 @@ elif which[0] == "numeric":
     nomask = None
     getmaskorNone = getmask
 elif which[0] == "numpy":
-    from numpy.core.ma import *
+    from maskedarray import *
     def getmaskorNone(obj):
         _msk = getmask(obj)
         if _msk is nomask:
