--- a/scipy/sparse/linalg/_dsolve/SuperLU/SRC/slu_Cnames.h
+++ b/scipy/sparse/linalg/_dsolve/SuperLU/SRC/slu_Cnames.h
@@ -29,6 +29,7 @@
 #ifndef __SUPERLU_CNAMES /* allow multiple inclusions */
 #define __SUPERLU_CNAMES
 
+#include "scipy_slu_config.h"
 
 #define ADD_       0
 #define ADD__      1
--- a/scipy/sparse/linalg/_dsolve/SuperLU/SRC/slu_util.h
+++ b/scipy/sparse/linalg/_dsolve/SuperLU/SRC/slu_util.h
@@ -34,6 +34,7 @@
 #include <assert.h>
 #include "superlu_enum_consts.h"
 
+#include "scipy_slu_config.h"
 
 /***********************************************************************
  * Macros
