diff --git a/scipy/_lib/_uarray/__init__.py b/scipy/_lib/_uarray/__init__.py
index 4d206b1c8..c7aa4ad6e 100644
--- a/scipy/_lib/_uarray/__init__.py
+++ b/scipy/_lib/_uarray/__init__.py
@@ -111,7 +111,5 @@ possible.

 from ._backend import *
 from ._backend import _Function
-from ._version import get_versions  # type: ignore

-__version__ = get_versions()["version"]
-del get_versions
+__version__ = 'v0.4+336.g4c712ad.scipy'
